Framework/Nexus/src/NeXusFile.cpp

```diff
@@ -74,165 +205,316 @@ template <> MANTID_NEXUS_DLL NXnumtype getType(char const) { return NXnumtype::C
+    : H5File(*pf), m_filename(pf->m_filename), m_access(pf->m_access), m_path(pf->m_path), m_current(pf->m_current),
+      m_close_handle(false), m_descriptor(pf->m_descriptor) {}
 
 File &File::operator=(File const &f) {
   if (this == &f) {
   } else {
     this->m_filename = f.m_filename;
     this->m_access = f.m_access;
-    this->m_pfile_id = f.m_pfile_id;
+    this->m_path = f.m_path;
+    this->m_current = f.m_current;
     this->m_close_handle = f.m_close_handle;
+    this->m_descriptor = f.m_descriptor;
   }
   return *this;
 }
 
-void File::initOpenFile(const string &filename, const NXaccess access) {
-  if (filename.empty()) {
-    throw Exception("Filename specified is empty constructor", "");
+// deconstructor
+
+File::~File() {
+  if (this->m_close_handle) {
+    H5::H5File::close();
   }
+}
 
-  NXhandle temp;
-  NXstatus status = NXopen(filename.c_str(), access, &(temp));
-  if (status != NXstatus::NX_OK) {
-    stringstream msg;
-    msg << "NXopen(" << filename << ", " << access << ") failed";
-    throw Exception(msg.str(), filename);
+void File::close() { H5::H5File::close(); }
+
+void File::flush(H5F_scope_t scope) const { H5::H5File::flush(scope); }
+
+H5::H5File File::getRoot() { return *this; }
+
+namespace {
+// methods for better exception messages
+template <typename T> std::string H5TypeAsString() { return "unknown"; }
+template <> std::string H5TypeAsString<H5::H5Object>() { return "H5Object"; }
+template <> std::string H5TypeAsString<H5::Group>() { return "Group"; }
+template <> std::string H5TypeAsString<H5::DataSet>() { return "DataSet"; }
+template <> std::string H5TypeAsString<H5::H5File>() { return "H5File"; }
+} // namespace
+
+/** Return a pointer corresponding to current location in file stack,
+ * cast to pointer of indicated type.
+ */
+template <typename T> std::shared_ptr<T> File::getCurrentLocationAs() {
+  std::shared_ptr<T> loc;
+  if (m_path == "/") {
+    std::shared_ptr<H5::H5File> proot = std::make_shared<H5::H5File>(getRoot().getId());
+    loc = std::dynamic_pointer_cast<T>(proot);
   } else {
-    this->m_pfile_id = std::make_shared<NXhandle>(temp);
+    loc = std::dynamic_pointer_cast<T>(m_current);
+  }
+  if (loc.get() == nullptr) {
+    std::stringstream msg;
+    msg << "Could not cast current location to needed H5Cpp type (requested=" << H5TypeAsString<T>() << ") at "
+        << getPath();
+    throw NXEXCEPTION(msg.str());
   }
+  return loc;
 }
 
-File::~File() {
-  if (m_close_handle && m_pfile_id != NULL) {
-    NXstatus status = NXclose(&(*this->m_pfile_id));
-    this->m_pfile_id = NULL;
-    if (status != NXstatus::NX_OK) {
-      stringstream msg;
-      msg << "NXclose failed for file " << m_filename << "\n";
-      NXReportError(const_cast<char *>(msg.str().c_str()));
+bool File::hasPath(std::string const &name) {
+  const std::string path = formAbsolutePath(name);
+  return m_descriptor.isEntry(path);
```

Comment:
Most of the problems I've seen in this file to compile in windows are related to interconversion between std::filesystem::path and std::string. This is prevalent in many places throughout the file. 
```suggestion
  const auto path = formAbsolutePath(name);
  return m_descriptor.isEntry(path.string());
```